src/Microsoft.Maintainability.Analyzers/Core/AvoidUninstantiatedInternalClasses.cs

```diff
@@ -37,6 +43,205 @@ public override void Initialize(AnalysisContext analysisContext)
+            analysisContext.RegisterCompilationStartAction(startContext =>
+            {
+                var instantiatedTypes = new List<INamedTypeSymbol>();
+                var internalTypes = new List<INamedTypeSymbol>();
+
+                // If the assembly being built by this compilation exposes its internals to
+                // any other assembly, don't report any "uninstantiated internal class" errors.
+                // If we were to report an error for an internal type that is not instantiated
+                // by this assembly, and then it turned out that the friend assembly did
+                // instantiate the type, that would be a false positive. We've decided it's
+                // better to have false negatives (which would happen if the type were *not*
+                // instantiated by any friend assembly, but we didn't report the issue) than
+                // to have false positives.
+                if (AssemblyExposesInternals(startContext.Compilation))
+                {
+                    return;
+                }
+
+                startContext.RegisterOperationAction(context =>
+                {
+                    IObjectCreationExpression expr = (IObjectCreationExpression)context.Operation;
+                    var namedType = expr.Type as INamedTypeSymbol;
+                    if (namedType != null)
+                    {
+                        instantiatedTypes.Add(namedType);
+                    }
+                }, OperationKind.ObjectCreationExpression);
+
+                startContext.RegisterSymbolAction(context =>
+                {
+                    INamedTypeSymbol type = (INamedTypeSymbol)context.Symbol;
+                    if (type.GetResultantVisibility() != SymbolVisibility.Public && !IsOkToBeUnused(type, context.Compilation))
+                    {
+                        internalTypes.Add(type);
+                    }
+                }, SymbolKind.NamedType);
+
+                startContext.RegisterCompilationEndAction(context =>
+                {
+                    IEnumerable<INamedTypeSymbol> uninstantiatedInternalTypes = internalTypes
+                        .Except(instantiatedTypes)
+                        .Where(type => !HasInstantiatedNestedType(type, instantiatedTypes));
+
+                    foreach (INamedTypeSymbol type in uninstantiatedInternalTypes)
+                    {
+                        context.ReportDiagnostic(type.CreateDiagnostic(Rule, type.FormatMemberName()));
+                    }
+                });
+            });
+        }
+
+        private bool AssemblyExposesInternals(Compilation compilation)
+        {
+            ISymbol assemblySymbol = compilation.Assembly;
+            ImmutableArray<AttributeData> attributes = assemblySymbol.GetAttributes();
+            return attributes.Any(
+                attr => attr.AttributeClass.Name.Equals("InternalsVisibleToAttribute", StringComparison.Ordinal));
```

Comment:
We want to match against fully qualified name - basically define a well known attribute type like [this](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Analyzer.Utilities/WellKnownTypes.cs#L199), and just check if attr.AttributeClass equals that symbol.